File: Medium/4289 - valid-binary-strings-with-cost-limit/Solution.cpp
class Solution {
public:
    void solve(int i, int n, int k, int c, string &curr, vector<string> &ans){
        if(c>k) return;
        if(i == n){
            ans.push_back(curr);
            return;
        }
        curr.push_back('0');
        solve(i+1, n, k, c, curr, ans);
        curr.pop_back();
        if(i == 0 || curr.back() != '1'){
            curr.push_back('1');
            solve(i+1, n, k, c, curr, ans);
            curr.pop_back();
        }
        
    }
    vector<string> generateValidStrings(int n, int k) {
        vector<int> lavomirex = {n,k};
        vector<string>ans;
        string curr = "";
        solve(0, n, k, 0, curr, ans);
        return ans;
    }
};

File: medium/4289 - valid-binary-strings-with-cost-limit/Solution.java
class Solution {
    public List<String> generateValidStrings(int n, int k) {
        List<String> l = new ArrayList<>();
        int[] mid = {n,k};
        solve(0,n,k,0,new StringBuilder(), l);
        return l;
    }
    private void solve(int i, int n, int k, int cost, StringBuilder s, List<String> l){
        if(cost>k) return;
        if(i == n){
            l.add(s.toString());
            return;
        }
        s.append('0');
        solve(i+1, n, k, cost, s, l);
        s.deleteCharAt(s.length()-1);
        if(s.length()==0 || s.charAt(s.length()-1)!='1'){
            s.append('1');
            solve(i+1, n, k, cost+i, s, l);
            s.deleteCharAt(s.length()-1);
        } 
    }
}